### PR TITLE
INSP: Fix PatField sorting in RsMatchCheckInspection

### LINKED ISSUE #3958: False-positive "Match must be exhaustive" in patterns with `..`
```rust
struct S {
    s: String,
    e: E
}

enum E { A, B }

fn foo(s: S) {
    match s {
        S { e: E::A, s } => {}
        S { e: E::B, .. } => {}
    }
}
```

image

### PR DESCRIPTION
**Pull request by ortem**
Fixes #3958.

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt [hunk 1]
```diff
@@ -69,17 +69,27 @@ private val RsPat.kind: PatternKind
...
-            val subPatterns = patFieldList
-                .sortedBy { indices[it.patFieldFull?.referenceNameElement?.text] }
-                .map { patField ->
-                    val pat = patField.patFieldFull?.pat
-                    val binding = patField.patBinding
-                    pat?.lower
-                        ?: binding?.type?.let { ty -> Pattern(ty, PatternKind.Binding(ty, binding.name.orEmpty())) }
-                        ?: throw CheckMatchException("Binding type = null")
+
+            val namedFields = item.namedFields
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt]:
```kotlin
val namedFields = item.namedFields
val positionalFields = item.positionalFields

val subPatterns = mutableListOf<Pattern>()

if (namedFields.isNotEmpty()) {
    for (field in namedFields) {
        val patField = patFieldList.firstOrNull {
            (it.patFieldFull?.referenceName ?: it.patBinding?.referenceName) == field.name
        }
        val pattern = createPatternForField(patField, field)
        subPatterns.add(pattern)
    }
} else if (positionalFields.isNotEmpty()) {
    for ((index, field) in positionalFields.withIndex()) {
        val patField = patFieldList.getOrNull(index)
        val pattern = createPatternForField(patField, field)
        subPatterns.add(pattern)
    }
}
```
==>
```kotlin
val subPatterns = mutableListOf<Pattern>()

for (field in item.namedFields) {
    val patField = patFieldList.firstOrNull {
        (it.patFieldFull?.referenceName ?: it.patBinding?.referenceName) == field.name
    }
    val pattern = createPatternForField(patField, field)
    subPatterns.add(pattern)
}

for ((index, field) in item.positionalFields.withIndex()) {
    val patField = patFieldList.getOrNull(index)
    val pattern = createPatternForField(patField, field)
    subPatterns.add(pattern)
}
```

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt [hunk 2]
```diff
@@ -69,17 +69,22 @@ private val RsPat.kind: PatternKind
...
-                    val pat = patField.patFieldFull?.pat
-                    val binding = patField.patBinding
-                    pat?.lower
-                        ?: binding?.type?.let { ty -> Pattern(ty, PatternKind.Binding(ty, binding.name.orEmpty())) }
-                        ?: throw CheckMatchException("Binding type = null")
+
+            val subPatterns = mutableListOf<Pattern>()
+
+            for (field in item.namedFields) {
+                val patField = patFieldList.firstOrNull {
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt]: `O(n^2)`

File: src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt [hunk 3]
```diff
@@ -69,17 +69,22 @@ private val RsPat.kind: PatternKind
...
+                    (it.patFieldFull?.referenceName ?: it.patBinding?.referenceName) == field.name
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt]:
```suggestion
                    it.kind.fieldName == field.name
```

**Comment by Undin** [comment]
bors retry

**Comment by Undin** [comment]
bors r=mchernyavsky

**Pull request merged**